assets/js/index.js

```diff
@@ -0,0 +1,97 @@
+let library = setupLibrary();
+
+if (!localStorage.uniqueId) {
+  localStorage.uniqueId = 1; 
+}
+
+function setupLibrary() {
+  let library;
+  if(localStorage.library) {
+    library = JSON.parse(localStorage.library).map((e) => Object.assign(new Book(), e));
+  } else {
+    library = [];
+  }
+  return library;
+}
+
+function addBookToLibrary() {
+  const form   = document.getElementById("book-form");
+
+  const title  = document.getElementById("title").value;
+  const author = document.getElementById("author").value;
+  const pages  = document.getElementById("pages").value;
+  const read   = document.getElementById("read").checked;
+
+  const book = new Book(title, author, pages, read);
+  library.push(book);
+  updateLocalStorage("library", library);
+  addBookToShelf(book);
+}
+
+function addBookToShelf(book) {
+  const table = document.getElementById("library-body");
+  const newRow = table.insertRow(table.rows.length);
+  newRow.setAttribute("id", "book-" + book.id);
+  newRow.innerHTML = book.render();
+}
+
+function removeBook(id) {
+  Swal.fire({
+    title: 'Are you sure?',
+    text: "You won't be able to revert this!",
+    type: 'warning',
+    showCancelButton: true,
+    confirmButtonColor: '#3085d6',
+    cancelButtonColor: '#d33',
+    confirmButtonText: 'Yes, delete it!'
+  }).then((result) => {
+    if (result.value) {
+      const book = library.find( (e) => e.id == id );
+      library.splice(library.indexOf(book), 1);
```

Comment:
Let's practice writing pure functions. Using `slice` here mutates a global variable. Now I know that you are intending to indeed have a new `library` which the book in question has already been removed. We can do the same like this:

```javascript
const newLibrary = library.filter(book => book.di ! == id)
updateLocalStorage("library", newLibrary)
```
Besides, it's much easier to do `filter` than `slice`. 

Since we are **practising** (darn Github wants me to use Canadian spelling!) functional programming with JavaScript, we have to get used to writing [pure functions](https://medium.com/@jamesjefferyuk/javascript-what-are-pure-functions-4d4d5392d49c).

Reply:
We tried to use pure functions on here, thanks for the article that is really good. But in here we cannot use the pure functions in here, cause the related function manipulates DOM and the local library constant must be changed cause our business logic still have to be inside the front-end. It's not a good practice to implement business logic inside front-end.

Reply:
Just to make it crystal clear, we have two options here: we can filter and re-assign library, but then library can't be a constant as you asked, or we can splice. You find filtering easier, I find splicing easier, but it makes no difference at the end of the day. Now, imagine the following situation: 
- library has books A, B and C
- you remove book A, library is filtered, local storage is updated with book B and C
- you remove book B, library is filtered, local storage is updated with book A (!) and C
- you refresh the page and... wtf? what is book A still doing here?

Reply:
Ok @madcido. I got your point man. In the second bullet point probably that's where our ideas differ. You seem to be starting with the 'old' library where you still had all three books A, B and C. I was coming from a perspective that your library array that you use in all your code is always the updated one (from `localStorage`). In that case if we purse my suggestion it would indeed cause you to re-architecture everything. But ok since you explained your idea very well and it's clearly just a difference in our perspectives, I'd say Im cool with it! :)